src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs

```diff
@@ -135,93 +135,174 @@ private void LoadTestsFromAnExtension(string extensionAssembly, IEnumerable<stri
-                    }
-
-                    // Collecting Data Point for Time Taken to Discover Tests by each Adapter
-                    this.requestData.MetricsCollection.Add(string.Format("{0}.{1}", TelemetryDataConstants.TimeTakenToDiscoverTestsByAnAdapter, discoverer.Metadata.DefaultExecutorUri), totalAdapterRunTime.TotalSeconds);
-                    totalTimeTakenByAdapters += totalAdapterRunTime.TotalSeconds;
+                    logger.SendMessage(TestMessageLevel.Warning,
+                        string.Format(CultureInfo.CurrentCulture, CrossPlatEngineResources.DeprecatedAdapterPath));
                 }
-                catch (Exception e)
+
+                // Collecting Data Point for Time Taken to Discover Tests by each Adapter
+                this.requestData.MetricsCollection.Add(
+                    string.Format("{0}.{1}", TelemetryDataConstants.TimeTakenToDiscoverTestsByAnAdapter,
+                        discoverer.Metadata.DefaultExecutorUri), totalAdapterRunTime.TotalSeconds);
+                totalTimeTakenByAdapters += totalAdapterRunTime.TotalSeconds;
+            }
+            catch (Exception e)
+            {
+                var message = string.Format(
+                    CultureInfo.CurrentUICulture,
+                    CrossPlatEngineResources.ExceptionFromLoadTests,
+                    discovererType.Name,
+                    e.Message);
+
+                logger.SendMessage(TestMessageLevel.Error, message);
+                EqtTrace.Error("DiscovererEnumerator.DiscoverTestsFromSingleDiscoverer: {0} ", e);
+            }
+        }
+
+        private static bool TryToLoadDiscoverer(LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities> discoverer, IMessageLogger logger, out Type discovererType)
+        {
+            discovererType = null;
+
+            // See if discoverer can be instantiated successfully else move next.
+            try
+            {
+                discovererType = discoverer.Value.GetType();
+            }
+            catch (Exception e)
+            {
+                var mesage = string.Format(
+                    CultureInfo.CurrentUICulture,
+                    CrossPlatEngineResources.DiscovererInstantiationException,
+                    e.Message);
+                logger.SendMessage(TestMessageLevel.Warning, mesage);
+                EqtTrace.Error("DiscovererEnumerator.LoadTestsFromAnExtension: {0} ", e);
+
+                return false;
+            }
+
+            return true;
+        }
+
+        private void RecordTotalTestsDiscoveredByCurrentAdapter(long currentTotalTests, LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities> discoverer,
+            ref bool discoverersFromDeprecatedLocations, ref double totalAdaptersUsed)
+        {
+            if (this.discoveryResultCache.TotalDiscoveredTests > currentTotalTests)
```

Comment:
we can inverse the if condition to reduce nesting.

Reply:
This is not required after removing over use of ref.